pkg/build/strategies/sti/sti.go

```diff
@@ -591,16 +590,17 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 	go dockerpkg.StreamContainerIO(errReader, &errOutput, func(a ...interface{}) { glog.Info(a...) })
 
 	err := builder.docker.RunContainer(opts)
-	if util.IsTimeoutError(err) {
-		// Cancel waiting for source input if the container timeouts
-		wg.Done()
-	}
 	if e, ok := err.(errors.ContainerError); ok {
 		// even with deferred close above, close errReader now so we avoid data race condition on errOutput;
 		// closing will cause StreamContainerIO to exit, thus releasing the writer in the equation
 		errReader.Close()
 		return errors.NewContainerError(config.BuilderImage, e.ErrorCode, errOutput)
 	}
+	// Do not wait for source input if the container times out.
+	// FIXME: this potentially leaks a goroutine.
+	if !util.IsTimeoutError(err) {
+		wg.Wait()
+	}
```

Comment:
@bparees does this make sense to you after all?

In my mind we're trading a panic for a leak, and based on previous discussions this is a "less critical" leak because this code is running in the builder pod and the process is short-lived.